### ISSUE TITLE
erratic behavior when typing in editor

### ISSUE THREAD

**Issue by aferriss** [processing/p5.js-web-editor#459]
#### Nature of issue?

- [x] Found a bug

#### Details about the bug: 

Every now and then when I'm typing, (especially after hitting backspace or spacebar) the cursor will jump a few lines and insert some random text. This is a hard bug to replicate especially because I can't reliably figure out what triggers it. 

FWIW, a co-teacher mentioned that he was seeing the same behavior, but only in Safari. I think I've seen it in all browsers, but I'm not 100% sure of that. 

- Web browser and version:  Chrome, Safari, Firefox
- Operating System: OSX
- Steps to reproduce this: Type often

**Comment by catarak** [comment]
thanks for reporting this! maybe i've never had this problem because i am not a very fast typer 🙃 i'll see if i can reproduce! i do most of the development on Chrome but i can try giving other browsers some TLC

**Comment by AidanNelson** [comment]
Hi!  I am also having this issue (on Chrome) and have noticed it especially when making quotation marks, though perhaps not exclusively then...  attached is a gif.  Thanks!
![gif](https://user-images.githubusercontent.com/6486359/32669330-2221b152-c60e-11e7-90fb-e44bbff74cf7.gif)

**Comment by catarak** [comment]
i've been trying to reproduce this but i haven't had any luck so far unfortunately! let me know if you figure out anymore specifics.

**Comment by aferriss** [comment]
Ok, I finally found something that reliably has weirdness. I'm testing on chrome and when I insert two spaces in between two close parens )) , it inserts a period. 

So if I have `abs(ceil(0.5));` and then I go insert spaces to be like `abs(ceil(0.5)  );` the editor will produce `abs(ceil(0.5) . );`

Seems to happen for me no matter the circumstance but only on closing parens. I also noticed this happens with square, angle, and curly brackets. so `[[]]` going to `[[]  ]` will produce `[[] . ]` 

Weirdly this happens in the github text editor too so maybe it's a browser issue.

I tested in firefox and it doesn't occur, but it does happen with safari, except it will create `((). )` instead of chrome's `(() . )`

**Comment by aferriss** [comment]
Also not sure if this is related, but I'm getting syntax errors where the shouldn't be any due to extra spaces. Take a look at this code

http://alpha.editor.p5js.org/aferriss/sketches/Sy_hPV_1f

The bug hint seems to thing there is an underscore in there somewhere.

**Comment by catarak** [comment]
if it happens on github too maybe it's an issue with CodeMirror? i can try updating CodeMirror to the latest release and maybe that will fix it.

**Comment by aferriss** [comment]
I did some digging and there's also a few closed issues on the codemirror repo that allude to strange behavior of cursors stemming from not calling the refresh function. I have definitely seen similar behavior to issue 5050 with the jumping cursor and text insertion in the wrong place.

https://github.com/codemirror/CodeMirror/issues/5040
https://github.com/codemirror/CodeMirror/issues/4997

I think I got to the bottom of the inserted period at least. I realized this happens even if you type into the url bar in chrome. Then I tested in Spotlight and got the same thing. So looks like this is something apple enabled by default. If you go to System Prefs > keyboard > text there is an option to turn this off. Still, I wonder if there is a way to short circuit this behavior. My feeling is that this shouldn't happen in a code editor (it doesn't in Sublime or XCode).

![screen shot 2017-11-14 at 9 15 26 am](https://user-images.githubusercontent.com/3698659/32794084-b9b508be-c91c-11e7-8d3e-3227dbea772f.png)
